network/peer/peer_test.go

```diff
@@ -273,3 +274,131 @@ func TestSend(t *testing.T) {
 	err = peer1.AwaitClosed(context.Background())
 	require.NoError(err)
 }
+
+func TestPingUptimes(t *testing.T) {
+	trackedSubnetID := ids.GenerateTestID()
+	untrackedSubnetID := ids.GenerateTestID()
+
+	trackedSubnets := set.NewSet[ids.ID](1)
+	trackedSubnets.Add(trackedSubnetID)
+
+	mc := newMessageCreator(t)
+
+	testCases := []struct {
+		name        string
+		msg         message.OutboundMessage
+		shouldClose bool
+		assertFn    func(*require.Assertions, *testPeer)
+	}{
+		{
+			name: "primary network only",
+			msg: func() message.OutboundMessage {
+				pingMsg, err := mc.Ping(1, nil)
+				require.NoError(t, err)
+				return pingMsg
+			}(),
+			assertFn: func(require *require.Assertions, peer *testPeer) {
+				uptime, ok := peer.ObservedUptime(constants.PrimaryNetworkID)
+				require.True(ok)
+				require.Equal(uint32(1), uptime)
+
+				uptime, ok = peer.ObservedUptime(trackedSubnetID)
+				require.False(ok)
+				require.Zero(uptime)
+			},
+		},
+		{
+			name: "primary network and subnet",
+			msg: func() message.OutboundMessage {
+				pingMsg, err := mc.Ping(
+					1,
+					[]*p2p.SubnetUptime{
+						{
+							SubnetId: trackedSubnetID[:],
+							Uptime:   1,
+						},
+					},
+				)
+				require.NoError(t, err)
+				return pingMsg
+			}(),
+			assertFn: func(require *require.Assertions, peer *testPeer) {
+				uptime, ok := peer.ObservedUptime(constants.PrimaryNetworkID)
+				require.True(ok)
+				require.Equal(uint32(1), uptime)
+
+				uptime, ok = peer.ObservedUptime(trackedSubnetID)
+				require.True(ok)
+				require.Equal(uint32(1), uptime)
+			},
+		},
+		{
+			name: "primary network and non tracked subnet",
+			msg: func() message.OutboundMessage {
+				pingMsg, err := mc.Ping(
+					1,
+					[]*p2p.SubnetUptime{
+						{
+							SubnetId: untrackedSubnetID[:],
+							Uptime:   1,
+						},
+						{
+							SubnetId: trackedSubnetID[:],
+							Uptime:   1,
+						},
+					},
+				)
+				require.NoError(t, err)
+				return pingMsg
+			}(),
+			shouldClose: true,
```

Comment:
Note that this test doesn't `assertFn` anything, because the peer should be closing... so the state doesn't really matter.